modules/tensor_mechanics/src/userobjects/nodal_patch_recovery/NPRBase.C

```diff
@@ -0,0 +1,173 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "NPRBase.h"
+#include "MathUtils.h"
+
+#include <Eigen/Dense>
+
+// TIMPI includes
+#include "timpi/communicator.h"
+#include "timpi/parallel_sync.h"
+
+InputParameters
+NPRBase::validParams()
+{
+  InputParameters params = ElementUserObject::validParams();
+
+  MooseEnum orders("CONSTANT FIRST SECOND THIRD FOURTH");
+  params.addRequiredParam<MooseEnum>(
+      "patch_polynomial_order",
+      orders,
+      "Polynomial order used in least squares fitting of material property "
+      "over the local patch of elements connected to a given node");
+
+  params.addRelationshipManager("ElementSideNeighborLayers",
+                                Moose::RelationshipManagerType::ALGEBRAIC,
+                                [](const InputParameters &, InputParameters & rm_params) {
+                                  rm_params.set<unsigned short>("layers") = 2;
+                                });
+
+  params.addParamNamesToGroup("patch_polynomial_order", "Advanced");
+
+  return params;
+}
+
+NPRBase::NPRBase(const InputParameters & parameters)
+  : ElementUserObject(parameters),
+    _qp(0),
+    _patch_polynomial_order(
+        static_cast<unsigned int>(getParam<MooseEnum>("patch_polynomial_order"))),
+    _multi_index(MathUtils::multiIndex(_mesh.dimension(), _patch_polynomial_order)),
+    _q(_multi_index.size())
+{
+}
+
+Real
+NPRBase::nodalPatchRecovery(const Point & x, const std::vector<dof_id_type> & elem_ids) const
+{
+  // before we go, check if we have enough sample points for solving the least square fitting
+  if (_q_point.size() * elem_ids.size() < _q)
+    mooseError("There are not enough sample points to recover the nodal value, try reducing the "
+               "polynomial order or using a higher-order quadrature scheme.");
+
+  // Assemble the least squares problem over the patch
+  RealEigenMatrix A = RealEigenMatrix::Zero(_q, _q);
+  RealEigenVector b = RealEigenVector::Zero(_q);
+  for (auto elem_id : elem_ids)
+  {
+    A += _Ae.at(elem_id);
+    b += _be.at(elem_id);
+  }
+
+  // Solve the least squares fitting
+  RealEigenVector coef = A.completeOrthogonalDecomposition().solve(b);
+
+  // Compute the fitted nodal value
+  RealEigenVector p = evaluateBasisFunctions(x);
+  return p.dot(coef);
+}
+
+RealEigenVector
+NPRBase::evaluateBasisFunctions(const Point & q_point) const
+{
+  RealEigenVector p(_q);
+  Real polynomial;
+  for (unsigned int r = 0; r < _multi_index.size(); r++)
```

Comment:
Or even better use 
```suggestion
  for (const auto & index: _multi_index)
```
here